x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java

```diff
@@ -91,13 +99,22 @@ public static void validate(BytesReference source) {
         }
     }
 
+    private static BytesReference replaceVariables(BytesReference input, String version, String versionProperty,
+                                                         Map<String, String> variables) {
+        String template = replaceVariable(input.utf8ToString(), versionProperty, version);
+        for (Map.Entry<String, String> variable : variables.entrySet()) {
+            template = replaceVariable(template, variable.getKey(), variable.getValue());
+        }
+        return new BytesArray(template);
+    }
+
     /**
-     * Filters the source: replaces any template version property with the version number
+     * Replaces all occurences of given variable with the value
      */
-    public static String filter(BytesReference source, String version, String versionProperty) {
-        return Pattern.compile(versionProperty)
-                .matcher(source.utf8ToString())
-                .replaceAll(version);
+    public static String replaceVariable(String input, String variable, String value) {
+        return Pattern.compile("${" + variable + "}", Pattern.LITERAL)
```

Comment:
it looks unsafe to inject a string into a regex pattern like this, it would be good to use `Pattern.quote(variable)` or disallow special characters

Reply:
That should be handled by the `LITERAL` flag, no?